File: Matrix.java
import java.util.Scanner;

public class Matrix {

    public static void main(String[] args) {
        int op, lin, col;
        Scanner in = new Scanner(System.in);
        System.out.println("1. Multiplication");
        System.out.println("2. Line Multiplication");
        System.out.println("3. Block Multiplication");
        System.out.print("Selection?: ");
        op = in.nextInt();
        System.out.print("Dimensions: lins=cols ? ");
        lin = in.nextInt();
        col = lin;
        double[][] testMat = createTestMatrixes(lin, col);

        double[] phc = new double[lin * col];
        long timeElapsed = 0;
        switch (op) {
            case 1:
                timeElapsed = onMult(phc, testMat[0], testMat[1], lin, col);
                break;
            case 2:
                timeElapsed = onMultLine(phc, testMat[0], testMat[1], lin, col);
                break;
            case 3:
                System.out.print("Block Size? ");
                onMultBlock(phc, testMat[0], testMat[1], lin, col, in.nextInt());
                break;
            default:
                System.out.println("Invalid select");
                in.close();
                return;
        }
        System.out.println("Time: " + (double) timeElapsed / 1000000000 + " seconds");
        displayFirstTen(phc, col);

        in.close();
    }

    static long onMult(double[] phc, double[] pha, double[] phb, int m_ar, int m_br) {
        long start = System.nanoTime();
        for (int i = 0; i < m_ar; i++) {
            for (int j = 0; j < m_br; j++) {
                int temp = 0;
                for (int k = 0; k < m_ar; k++) {
                    temp += pha[i * m_ar + k] * phb[k * m_br + j];
                }
                phc[i * m_ar + j] = temp;
            }
        }
        long finish = System.nanoTime();
        return finish - start;
    }

    static long onMultLine(double[] phc, double[] pha, double[] phb, int m_ar, int m_br) {
        long start = System.nanoTime();
        for (int i = 0; i < m_ar; i++) {
            for (int k = 0; k < m_ar; k++) {
                for (int j = 0; j < m_br; j++) {
                    phc[i * m_ar + j] += pha[i * m_ar + k] * phb[k * m_br + j];
                }
            }
        }
        long finish = System.nanoTime();
        return finish - start;
    }

    static long onMultBlock(double[] phc, double[] pha, double[] phb, int m_ar, int m_br, int bkSize) {
        long start = System.nanoTime();


        for (int bi = 0; bi < m_ar; bi += bkSize) {
            for (int bj = 0; bj < m_br; bj += bkSize) {
                for (int bk = 0; bk < m_ar; bk += bkSize) {
                    for (int i = 0; i < bkSize; i++) {
                        for (int j = 0; j < bkSize; j++) {
                            for (int k = 0; k < bkSize; k++) {
                                phc[(bi + i) * m_ar + bj + j] += pha[(bi + i) * m_ar + bk + k]
                                        * phb[(bk + k) * m_ar + bj + j];
                            }
                        }
                    }
                }
            }
        }

        long finish = System.nanoTime();
        return finish - start;
    }

    static double[][] createTestMatrixes(int m_ar, int m_br) {
        double[] pha = new double[m_ar * m_ar];
        double[] phb = new double[m_br * m_br];

        for (int i = 0; i < m_ar; i++)
            for (int j = 0; j < m_ar; j++)
                pha[i * m_ar + j] = (double) 1.0;

        for (int i = 0; i < m_br; i++)
            for (int j = 0; j < m_br; j++)
                phb[i * m_br + j] = (double) (i + 1);

        return new double[][] { pha, phb };
    }

    static void displayFirstTen(double[] phc, int m_br) {
        // display 10 elements of the result matrix tto verify correctness
        System.out.println("Result Matrix:");

        for (int i = 0; i < 1; i++) {
            for (int j = 0; j < Math.min(10, m_br); j++)
                System.out.print(phc[j] + " ");
        }
        System.out.println();
    }
}

File: matrix.cpp
#include <stdio.h>
#include <iostream>
#include <iomanip>
#include <time.h>
#include <cstdlib>
#include <string.h>
#include <papi.h>
#include <omp.h>

using namespace std;

#define SYSTEMTIME clock_t

void OnMult(int m_ar, int m_br, int flag_parallel)
{

    SYSTEMTIME Time1, Time2;

    char st[100];
    double temp;
    int i, j, k;

    double *pha, *phb, *phc;

    pha = (double *)malloc((m_ar * m_ar) * sizeof(double));
    phb = (double *)malloc((m_ar * m_ar) * sizeof(double));
    phc = (double *)malloc((m_ar * m_ar) * sizeof(double));

    bzero(pha, (m_ar * m_ar));
    bzero(phb, (m_ar * m_ar));
    bzero(phc, (m_ar * m_ar));

    for (i = 0; i < m_ar; i++)
        for (j = 0; j < m_ar; j++)
            pha[i * m_ar + j] = (double)1.0;

    for (i = 0; i < m_br; i++)
        for (j = 0; j < m_br; j++)
            phb[i * m_br + j] = (double)(i + 1);

    Time1 = clock();
    time_t seconds;
    seconds=time(NULL);
    if(flag_parallel){
        #pragma omp parallel private(i, j)
        for (i = 0; i < m_ar; i++)
        {
            for (j = 0; j < m_br; j++)
            {
                #pragma omp single
                temp = 0;
                #pragma omp for reduction(+:temp)
                for (k = 0; k < m_ar; k++)
                {
                    temp += pha[i * m_ar + k] * phb[k * m_br + j];
                }
                #pragma omp single
                phc[i * m_ar + j] = temp;
            }
        } 

    }

    else{
        for (i = 0; i < m_ar; i++)
        {
            for (j = 0; j < m_br; j++)
            {
                temp = 0;
                for (k = 0; k < m_ar; k++)
                {
                    temp += pha[i * m_ar + k] * phb[k * m_br + j];
                }
                phc[i * m_ar + j] = temp;
            }
        }
    }

    time_t seconds2;
    seconds2=time(NULL);
    printf("Time2: %ld \n", seconds2-seconds);
    Time2 = clock();
    sprintf(st, "Time: %3.3f seconds\n", (double)(Time2 - Time1) / CLOCKS_PER_SEC);
    cout << st;

    // display 10 elements of the result matrix tto verify correctness
    cout << "Result matrix: " << endl;
    for (i = 0; i < 1; i++)
    {
        for (j = 0; j < min(10, m_br); j++)
            cout << phc[j] << " ";
    }
    cout << endl;

    free(pha);
    free(phb);
    free(phc);
}

// add code here for line x line matriz multiplication
void OnMultLine(int m_ar, int m_br, int flag_parallel)
{

    SYSTEMTIME Time1, Time2;

    char st[100];
    int i, j, k;

    double *pha, *phb, *phc;

    pha = (double *)malloc((m_ar * m_ar) * sizeof(double));
    phb = (double *)malloc((m_ar * m_ar) * sizeof(double));
    phc = (double *)malloc((m_ar * m_ar) * sizeof(double));

    bzero(pha, (m_ar * m_ar));
    bzero(phb, (m_ar * m_ar));
    bzero(phc, (m_ar * m_ar));

    for (i = 0; i < m_ar; i++)
        for (j = 0; j < m_ar; j++)
            pha[i * m_ar + j] = (double)1.0;

    for (i = 0; i < m_br; i++)
        for (j = 0; j < m_br; j++)
            phb[i * m_br + j] = (double)(i + 1);

    Time1 = clock();

    for (i = 0; i < m_ar; i++)
    {
        for (k = 0; k < m_ar; k++)
        {
#pragma omp parallel for
            for (j = 0; j < m_br; j++)
            {
                phc[i * m_ar + j] += pha[i * m_ar + k] * phb[k * m_br + j];
            }
        }
    }

    Time2 = clock();
    sprintf(st, "Time: %3.3f seconds\n", (double)(Time2 - Time1) / CLOCKS_PER_SEC);
    cout << st;

    // display 10 elements of the result matrix tto verify correctness
    cout << "Result matrix: " << endl;
    for (i = 0; i < 1; i++)
    {
        for (j = 0; j < min(10, m_br); j++)
            cout << phc[j] << " ";
    }
    cout << endl;

    free(pha);
    free(phb);
    free(phc);
}

// add code here for block x block matriz multiplication
void OnMultBlock(int m_ar, int m_br, int bkSize, int flag_parallel)
{

    SYSTEMTIME Time1, Time2;

    char st[100];
    int i, j, k;

    double *pha, *phb, *phc;

    pha = (double *)malloc((m_ar * m_ar) * sizeof(double));
    phb = (double *)malloc((m_ar * m_ar) * sizeof(double));
    phc = (double *)malloc((m_ar * m_ar) * sizeof(double));

    bzero(pha, (m_ar * m_ar));
    bzero(phb, (m_ar * m_ar));
    bzero(phc, (m_ar * m_ar));

    for (i = 0; i < m_ar; i++)
        for (j = 0; j < m_ar; j++)
            pha[i * m_ar + j] = (double)1.0;

    for (i = 0; i < m_br; i++)
        for (j = 0; j < m_br; j++)
            phb[i * m_br + j] = (double)(i + 1);

    Time1 = clock();

    for (int bi = 0; bi < m_ar; bi += bkSize)
    {
        for (int bj = 0; bj < m_br; bj += bkSize)
        {
            for (int bk = 0; bk < m_ar; bk += bkSize)
            {
                // Line mult
                for (i = 0; i < bkSize; i++)
                {
                    for (k = 0; k < bkSize; k++)
                    {
                        for (j = 0; j < bkSize; j++)
                        {
                            phc[(bi + i) * m_ar + bj + j] += pha[(bi + i) * m_ar + bk + k] * phb[(bk + k) * m_ar + bj + j];
                        }
                    }
                }
            }
        }
    }

    Time2 = clock();
    sprintf(st, "Time: %3.3f seconds\n", (double)(Time2 - Time1) / CLOCKS_PER_SEC);
    cout << st;

    // display 10 elements of the result matrix tto verify correctness
    cout << "Result matrix: " << endl;
    for (i = 0; i < 1; i++)
    {
        for (j = 0; j < min(10, m_br); j++)
            cout << phc[j] << " ";
    }
    cout << endl;

    free(pha);
    free(phb);
    free(phc);
}

void handle_error(int retval)
{
    printf("PAPI error %d: %s\n", retval, PAPI_strerror(retval));
    exit(1);
}

void init_papi()
{
    int retval = PAPI_library_init(PAPI_VER_CURRENT);
    if (retval != PAPI_VER_CURRENT && retval < 0)
    {
        printf("PAPI library version mismatch!\n");
        exit(1);
    }
    if (retval < 0)
        handle_error(retval);

    std::cout << "PAPI Version Number: MAJOR: " << PAPI_VERSION_MAJOR(retval)
              << " MINOR: " << PAPI_VERSION_MINOR(retval)
              << " REVISION: " << PAPI_VERSION_REVISION(retval) << "\n";
}

int main(int argc, char *argv[])
{

    int lin, col, blockSize, isParallel;
    int op;

    int EventSet = PAPI_NULL;
    long long values[2];
    int ret;

    ret = PAPI_library_init(PAPI_VER_CURRENT);
    if (ret != PAPI_VER_CURRENT)
        std::cout << "FAIL" << endl;

    ret = PAPI_create_eventset(&EventSet);
    if (ret != PAPI_OK)
        cout << "ERROR: create eventset" << endl;

    ret = PAPI_add_event(EventSet, PAPI_L1_DCM);
    if (ret != PAPI_OK)
        cout << "ERROR: PAPI_L1_DCM" << endl;

    ret = PAPI_add_event(EventSet, PAPI_L2_DCM);
    if (ret != PAPI_OK)
        cout << "ERROR: PAPI_L2_DCM" << endl;

    op = 1;
    do
    {
        cout << endl
             << "1. Multiplication" << endl;
        cout << "2. Line Multiplication" << endl;
        cout << "3. Block Multiplication" << endl;
        cout << "Selection?: ";
        cin >> op;
        if (op == 0)
            break;
        printf("Dimensions: lins=cols ? ");
        cin >> lin;
        col = lin;
        
        printf("Is parallel (0 or 1): ");
        cin >> isParallel;

        // Start counting
        ret = PAPI_start(EventSet);
        if (ret != PAPI_OK)
            cout << "ERROR: Start PAPI" << endl;

        switch (op)
        {
        case 1:
            OnMult(lin, col, isParallel);
            break;
        case 2:
            OnMultLine(lin, col, isParallel);
            break;
        case 3:
            cout << "Block Size? ";
            cin >> blockSize;
            OnMultBlock(lin, col, blockSize, isParallel);
            break;
        }

        ret = PAPI_stop(EventSet, values);
        if (ret != PAPI_OK)
            cout << "ERROR: Stop PAPI" << endl;
        printf("L1 DCM: %lld \n", values[0]);
        printf("L2 DCM: %lld \n", values[1]);

        ret = PAPI_reset(EventSet);
        if (ret != PAPI_OK)
            std::cout << "FAIL reset" << endl;

    } while (op != 0);

    ret = PAPI_remove_event(EventSet, PAPI_L1_DCM);
    if (ret != PAPI_OK)
        std::cout << "FAIL remove event" << endl;

    ret = PAPI_remove_event(EventSet, PAPI_L2_DCM);
    if (ret != PAPI_OK)
        std::cout << "FAIL remove event" << endl;

    ret = PAPI_destroy_eventset(&EventSet);
    if (ret != PAPI_OK)
        std::cout << "FAIL destroy" << endl;
}
